### PR TITLE
Unary negation for integers

### PR DESCRIPTION
**Pull request by zphixon**
Howdy! This is part of #1983, only looking at integers for now. E.g.

```gleam
let a = 6
let b = -a
```

now parses, type checks, and codegens, where previously `-a` would be a syntax error.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -49,6 +49,7 @@
   runtime if executed.
 - Fixed bug where raised errors on JS would have an extra stack frame recorded
   in them.
+- Added support for unary negation of integers.
```
Comment (lpil) [inline CHANGELOG.md:52]: Could you rebase on main and put this in a new `## Unreleased` section please?

File: compiler-core/src/ast/typed.rs [hunk 2]
```diff
@@ -152,7 +152,12 @@ pub enum TypedExpr {
         args: Vec<TypedRecordUpdateArg>,
     },
 
-    Negate {
+    NegateBool {
+        location: SrcSpan,
+        value: Box<Self>,
+    },
+
+    NegateInteger {
```
Comment (lpil) [inline compiler-core/src/ast/typed.rs:160]:
```suggestion
    NegateInt {
```

To be consistent within the codebase 🙏

File: compiler-core/src/ast/untyped.rs [hunk 3]
```diff
@@ -123,7 +123,12 @@ pub enum UntypedExpr {
         arguments: Vec<UntypedRecordUpdateArg>,
     },
 
-    Negate {
+    NegateBool {
+        location: SrcSpan,
+        value: Box<Self>,
+    },
+
+    NegateInteger {
```
Comment (lpil) [inline compiler-core/src/ast/untyped.rs:131]:
```suggestion
    NegateInt {
```

File: compiler-core/src/type_/expression.rs [hunk 4]
```diff
@@ -370,12 +374,31 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
...
         unify(bool(), value.type_()).map_err(|e| convert_unify_error(e, value.location()))?;
 
-        Ok(TypedExpr::Negate {
+        Ok(TypedExpr::NegateBool {
+            location,
+            value: Box::new(value),
+        })
+    }
+
+    fn infer_negate_integer(
```
Comment (lpil) [inline compiler-core/src/type_/expression.rs:392]:
```suggestion
    fn infer_negate_int(
```

File: compiler-core/src/format.rs [hunk 5]
```diff
@@ -692,7 +692,9 @@ impl<'comments> Formatter<'comments> {
 
             UntypedExpr::TupleIndex { tuple, index, .. } => self.tuple_index(tuple, *index),
 
-            UntypedExpr::Negate { value, .. } => self.negate(value),
+            UntypedExpr::NegateInteger { value, .. } => self.negate_with("-", value),
```
Comment (lpil) [inline compiler-core/src/format.rs:695]: We had someone the other day confused by the behaviour of the code `--x`, let's make it clear there's no `--` operator in Gleam by adding spaces or blocks

File: compiler-core/src/parse.rs [hunk 6]
```diff
@@ -577,7 +577,27 @@ where
...
+                    None => {
+                        return parse_error(
+                            ParseErrorType::ExpectedExpr,
+                            SrcSpan { start, end: start },
+                        )
+                    }
+                }
+            }
+
+            // Integer negation
```
Comment (lpil) [inline compiler-core/src/parse.rs:596]:
```suggestion
            // Int negation
```

**Review by lpil — COMMENTED** [review]
Very nice! Could we get some tests for the Erlang and JavaScript code generation, as well as for the formatter. Thanks!

File: compiler-core/src/format.rs (hunk 5 above)
Comment (inoas) [reply compiler-core/src/format.rs:695]: `NegateInteger`

Would `NegateInt` be consistent with the other change requests?
`Integer` currently only exists 5 times in the whole gleam-lang/gleam repo, `Int` 1161 times.

Comment (zphixon) [reply compiler-core/src/format.rs:695]: So

```gleam
let a = -b
```

should format to

```gleam
let a = - b
```

or just the case of repeated negation,

```gleam
let a = --b
```

to one of

```gleam
let a = - { -b }
let a = - -b
```

Comment (inoas) [reply compiler-core/src/format.rs]: I guess?:

- `let a = -b` is fine and expected.
- `let a = - b` should format to `let a = -b`
- `let a = --b` should format to `let a = b`

Comment (zphixon) [reply compiler-core/src/format.rs]: That's an interesting idea. Bool negation doesn't do that right now, maybe that could be a feature request?

I think blocks look the nicest, e.g. `--a` becomes `- { -a }`. The newest changes do that

Comment (lpil) [reply compiler-core/src/format.rs]: The formatter doesn't change the AST at all so we wouldn't collapse negation operators, even if they are doing something pointless.  Using a block sounds great

File: compiler-core/src/format.rs [hunk 7]
```diff
@@ -1529,13 +1531,24 @@ impl<'comments> Formatter<'comments> {
...
             _ => docvec!["!", self.wrap_expr(expr)],
         }
     }
 
+    fn negate_int<'a>(&mut self, expr: &'a UntypedExpr) -> Document<'a> {
+        match expr {
+            // Always nest repeated negation in a block to avoid confusion with
+            // the pre-decrement operator (which does not exist)
+            UntypedExpr::BinOp { .. } | UntypedExpr::NegateInt { .. } => {
+                docvec!["- { ", self.expr(expr), " }"]
```
Comment (lpil) [inline compiler-core/src/format.rs:1546]: This unfortunately won't wrap and indent correctly when the line is too long. Let's extract the block wrapping code from `wrap_expr` into a helper function and use that here.

Comment (zphixon) [reply compiler-core/src/format.rs]: I ended up cribbing from `operator_side`, `wrap_expr` includes some newlines that weren't always necessary.

**Review by lpil — APPROVED** [review]
Wonderful work! Thank you for being so thorough with the testing as well!

**Pull request merged**